kernel/src/fs/path/dentry.rs

```diff
@@ -261,13 +292,44 @@ impl Dentry {
             return_errno!(Errno::ENOTDIR);
         }
 
+        if is_dot(name) {
+            return_errno_with_message!(Errno::EINVAL, "rmdir on .");
+        }
+        if is_dotdot(name) {
+            return_errno_with_message!(Errno::ENOTEMPTY, "rmdir on ..");
+        }
+
         let children = self.children.upread();
         children.check_mountpoint(name)?;
 
-        self.inode.rmdir(name)?;
-
         let mut children = children.upgrade();
-        children.delete(name);
+        let cached_child = children.delete(name);
+
+        let child_inode = match cached_child {
+            Some(child) => {
+                // Cache hit: use the cached dentry
+                child.inode().clone()
+            }
+            None => {
+                // Cache miss: need to lookup from the underlying filesystem
+                drop(children);
+                self.inode.lookup(name)?
+            }
+        };
+
+        self.inode.rmdir(name)?;
+        let nlinks = child_inode.metadata().nlinks;
+        if nlinks == 0 {
+            fs::notify::on_inode_removed(&child_inode);
+        }
+        fs::notify::on_delete(self.inode(), &child_inode, String::from(name));
```

Comment:
This `on_delete` is not zero cost. Even both the dir and removed inodes are not monitored, we still need to pay the price of allocating the name string.